pymatchseries/implementation/implementation.py

```diff
@@ -0,0 +1,337 @@
+
+
+class JNRR:
+    def __init__(
+        self,
+        images: Signal2D,
+    ) -> None:
+        self._validate_images(images)
+        self.__config = JNNRConfig()
+        self.__state = JNNRState(images)
+
+    @classmethod
+    def load(
+        cls,
+        filepath: Path,
+    ) -> JNRR:
+        raise NotImplementedError()
+
+    def save(self) -> None:
+        raise NotImplementedError()
+
+    @property
+    def images(self) -> Signal2D:
+        return self.state.images
+
+    @property
+    def state(self) -> JNNRState:
+        return self.__state
+
+    def run(self) -> None:
+        L = self.config.regularization.constant_start
+        n_stages = self.config.n_stages
+
+        for stage in range(n_stages):
+            displacements = []
+            corrected_images = []
+
+            progress = tqdm(total=self.number_of_images)
+
+            with progress:
+                progress.set_description(
+                    f"Stage: {stage + 1}, Image: 0/{self.number_of_images}"
+                )
+                # Registration - finding all displacements
+                for i, image in enumerate(
+                    self._get_image_iterator(
+                        self.images,
+                        device=self.config.device,
+                    )
+                ):
+                    dp = get_dispatcher(image)
+                    if self.state.reference_image is None:
+                        self.state.reference_image = image
+                        displacement = dp.zeros((2, *image.shape), dtype=image.dtype)
+                        displacements.append(
+                            self._displacement_to_complex(displacement)
+                        )
+                        corrected_images.append(image)
+                        progress.update(n=1)
+                        continue
+
+                    displacement = self._get_multilevel_displacement_field(
+                        image,
+                        self.state.reference_image,
+                        regularization_constant=L,
+                        configuration=self.config,
+                    )
+                    displacements.append(self._displacement_to_complex(displacement))
+
+                    corrected_image = self._apply_displacement(image, displacement)
+                    corrected_images.append(corrected_image)
+
+                    progress.set_description(
+                        f"Stage: {stage + 1}, Image: {i + 1}/{self.number_of_images}"
+                    )
+                    progress.update(n=1)
+
+                # Bias correction??
+
+                self.state.reference_image = self._aggregate_stack(
+                    dp.stack(corrected_images),
+                )
+                self.state.deformations = ComplexSignal2D(dp.stack(corrected_images))
+                self.state.completed_stages = stage + 1
+                L *= self.config.regularization.factor_stage
```

Comment:
This seems slightly different from what my C++ implementation is doing. There, the first stage uses `lambda`, all other stages use `lambda*extraStagesLambdaFactor`, i.e. the regularization parameter is only changed when moving from stage 1 to stage 2. It's not changed further when going from stage 2 to 3..